esphome/components/sonoff_d1/sonoff_d1.cpp

```diff
@@ -0,0 +1,312 @@
+  uint8_t ref_buffer[7] = {0xAA, 0x55, cmd[2], cmd[3], 0x00, 0x00, 0x00};
+  uint8_t buffer[sizeof(ref_buffer)] = {0};
+  uint32_t pos = 0, buf_len = sizeof(ref_buffer);
+
+  // Update the reference checksum
+  this->populate_checksum_(ref_buffer, sizeof(ref_buffer));
+
+  // Read ack code, this either reads 7 bytes or exits with a timeout
+  this->read_command_(buffer, buf_len);
+
+  // Compare response with expected response
+  while (pos < sizeof(ref_buffer) && ref_buffer[pos] == buffer[pos]) {
+    pos++;
+  }
+  if (pos == sizeof(ref_buffer)) {
+    ESP_LOGD(TAG, "[%04d] Acknowledge received", this->write_count_);
+    return true;
+  } else {
+    ESP_LOGW(TAG, "[%04d] Unexpected acknowledge received (possible clash of RF/HA commands), expected ack was:",
+             this->write_count_);
+    ESP_LOGW(TAG, "[%04d] %s", this->write_count_, format_hex_pretty(ref_buffer, sizeof(ref_buffer)).c_str());
+  }
+  return false;
+}
+
+bool SonoffD1Output::write_command_(uint8_t *cmd, const size_t len, bool needs_ack) {
+  // Do some consistency checks
+  if (len < 7) {
+    ESP_LOGW(TAG, "[%04d] Too short command (actual len is %d bytes, minimal is 7)", this->write_count_, len);
+    return false;
+  }
+  if (cmd[0] != 0xAA || cmd[1] != 0x55) {
+    ESP_LOGW(TAG, "[%04d] Wrong header (%x%x, must be AA55)", this->write_count_, cmd[0], cmd[1]);
+    return false;
+  }
+  if ((cmd[5] + 7 /*mandatory header + suffix length*/) != len) {
+    ESP_LOGW(TAG, "[%04d] Payload length field does not match packet lenght (%d, expected %d)", this->write_count_,
+             cmd[5], len - 7);
+    return false;
+  }
+  this->populate_checksum_(cmd, len);
+
+  // Need retries here to handle the following cases:
+  // 1. On power up companion MCU starts to respond with a delay, so few first commands are ignored
+  // 2. UART command initiated by this component can clash with a command initiated by RF
+  uint32_t retries = 10;
+  do {
+    ESP_LOGV(TAG, "[%04d] Writing to the dimmer:", this->write_count_);
+    ESP_LOGV(TAG, "[%04d] %s", this->write_count_, format_hex_pretty(cmd, len).c_str());
+    this->write_array(cmd, len);
+    this->write_count_++;
+    if (!needs_ack)
+      return true;
+    retries--;
+  } while (!this->read_ack_(cmd, len) && retries > 0);
+
+  if (retries) {
+    return true;
+  } else {
+    ESP_LOGE(TAG, "[%04d] Unable to write to the dimmer", this->write_count_);
+  }
+  return false;
+}
+
+bool SonoffD1Output::control_dimmer_(const bool binary, const uint8_t brightness) {
+  // Include our basic code from the Tasmota project, thank you again!
+  //                    0     1     2     3     4     5     6     7     8
+  uint8_t cmd[17] = {0xAA, 0x55, 0x01, 0x04, 0x00, 0x0A, 0x00, 0x00, 0xFF,
+                     // 9     10    11    12    13    14    15    16
+                     0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x00};
+
+  cmd[6] = binary;
+  cmd[7] = this->remap_(brightness, 0, 100, this->min_value_, this->max_value_);
```

Comment:
This exists in `esphome/core/helpers.h` (It might not have when you created this code though)
```suggestion
  cmd[7] = remap(brightness, 0, 100, this->min_value_, this->max_value_);
```

Reply:
I cannot check this change now (travelling) but I suspect the first multiplicaion will cause integer overflow inside esphome/core/helpers.h as it will be done in uint8_t.